flamingpy/codes/graphs/egraph.py

```diff
@@ -214,3 +214,111 @@ def draw_adj(self, **kwargs):
 
         adj = self.adj_generator(sparse=False)
         return plot_mat_heat_map(adj, **kwargs)
+
+    def add_qubit(self, qubit=None, index=None, neighbors=None) -> None:
+        """Add qubit to EGraph at position with neigbours in
+        existing_neighbours.
+
+        Args:
+            qubit (3D tuple, or None): qubit to add. If it is a 3D tuple,
+            the qubit is added in that position. If qubit is None, the qubit is
+            positioned one unit further than the maximum position in the z
+            direction in position (0,0,z_max + 1).
+
+            index (int or None): index of qubit. If int, the qubit is
+            positioned at that index. If None, the index of the qubit is the
+            last one.
+
+            neighbours (list[int] or list[tuple]): neighbors of qubit specified
+            with indices or positions, respectively.
+        """
+        # Add qubit
+        if isinstance(qubit, tuple):
+            if len(qubit) != 3:
+                raise Exception(
+                    "The position should be a 3D tuple, but it was given a "
+                    + f"{len(qubit)}D tuple"
+                )
+        elif qubit is None:
+            z_max = max(map(lambda x: x[2], self.nodes()))
+            qubit = (0, 0, z_max + 1)
+        else:
+            raise Exception(
+                "Qubit type not supported. Excepted 3D tuple or None, but was"
+                + f" given {type(qubit)}"
+            )
+        self.add_node(qubit)
+
+        # Update Neighbors
+        if neighbors is not None:
+            if len(neighbors) != 0:
+                if isinstance(neighbors[0], int):
+                    neighbors = [(self.to_points[ind], qubit) for ind in neighbors.copy()]
+                elif isinstance(neighbors[0], tuple):
+                    neighbors = [(neigh, qubit) for neigh in neighbors.copy()]
+
+                self.add_edges_from(neighbors)
+
+        # Update dictionaries
+        if (self.to_points is not None) and (self.to_indices is not None):
+            self._update_dicts_add_qubit(qubit, index)
+
+        self.adj_mat = None
+
+    def _update_dicts_add_qubit(self, qubit, index):
+        """Update dictionaries of class when adding a qubit."""
+
+        if isinstance(index, int):
+            self.to_points = {k + 1 if k >= index else k: v for k, v in self.to_points.items()}
+            self.to_indices = {v: k for k, v in self.to_points.items()}
+            new_index = index
+        elif index is None:
+            new_index = max(self.to_points.keys()) + 1
+        else:
+            raise Exception(
+                "Index type not supported. Expected int or None, but was given" + f" {type(index)}"
+            )
+        self.to_points[new_index] = qubit
+        self.to_indices[qubit] = new_index
+
+    def remove_qubit(self, qubit=None) -> None:
```

Comment:
I think `qubit` should never be None -- we would like the user to be explicit about which qubit to remove.

Reply:
Makes sense, I've removed this